BUG: Fix using "inf"/"-inf" in na_values for csv with int index column

Templarrr opened the pull request:
- [x] closes #17128
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The issue happens when you try to use 'inf' or '-inf' as part of na_values in read_csv.

Code snippet to reproduce:
```
from StringIO import StringIO

import pandas as pd

dataset = StringIO('''index,col1,col2,col3
1,6,10,14
2,7,11,15
3,8,12,16
4,9,13,17
5,inf,-inf,bla
''')
na_values = ['inf', '-inf', 'bla']

df = pd.read_csv(dataset, na_values=na_values, index_col='index')
print df
print df.dtypes
```

Without fix:
```
Traceback (most recent call last):
  File "/home/modintsov/workspace/DataRobot/playground.py", line 39, in <module>
    df = pd.read_csv(dataset, na_values=na_values, index_col='index')
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 678, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 446, in _read
    data = parser.read(nrows)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1036, in read
    ret = self._engine.read(nrows)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1922, in read
    index, names = self._make_index(data, alldata, names)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1426, in _make_index
    index = self._agg_index(index)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1520, in _agg_index
    arr, _ = self._infer_types(arr, col_na_values | col_na_fvalues)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1600, in _infer_types
    mask = algorithms.isin(values, list(na_values))
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/core/algorithms.py", line 418, in isin
    values, _, _ = _ensure_data(values, dtype=dtype)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/core/algorithms.py", line 82, in _ensure_data
    return _ensure_int64(values), 'int64', 'int64'
  File "pandas/_libs/algos_common_helper.pxi", line 3227, in pandas._libs.algos.ensure_int64
  File "pandas/_libs/algos_common_helper.pxi", line 3232, in pandas._libs.algos.ensure_int64
OverflowError: cannot convert float infinity to integer
```

With fix (as expected):
```
       col1  col2  col3
index
1       6.0  10.0  14.0
2       7.0  11.0  15.0
3       8.0  12.0  16.0
4       9.0  13.0  17.0
5       NaN   NaN   NaN
col1    float64
col2    float64
col3    float64
dtype: object
```

Commit:
BUG: Fix using "inf"/"-inf" in na_values for csv with int index column

Commit:
Add the test case for inf as na

Commit:
Changelog

pep8speaks commented:
Hello @Templarrr! Thanks for submitting the PR.



 - In the file [`pandas/tests/io/parser/na_values.py`](https://github.com/pandas-dev/pandas/blob/65f83d6c779aefa153943b7e9849388c49550aa0/pandas/tests/io/parser/na_values.py), following are the PEP8 issues :

> [Line 377:80](https://github.com/pandas-dev/pandas/blob/65f83d6c779aefa153943b7e9849388c49550aa0/pandas/tests/io/parser/na_values.py#L377): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)

Commit:
Linter

jreback commented on doc/source/whatsnew/v0.23.4.txt:
```diff
@@ -41,7 +41,7 @@ Bug Fixes
 **Indexing**
 
--
+- Fix OverflowError when trying to use 'inf' as na_value with int index column
```
add the issue number here, use double-backticks around OverflowError. use double backticks on na_value; spell out int -> integer.

jreback commented on pandas/tests/io/parser/na_values.py:
```diff
@@ -369,3 +369,13 @@ def test_no_na_filter_on_index(self):
         tm.assert_frame_equal(out, expected)
+
+    def test_inf_na_values_with_int_index(self):
+        data = "idx,col1,col2\n1,3,4\n2,inf,-inf"
```
add the issue number here as a comment

Templarrr commented:
Oh, wow, it's actually my colleague :) 
Didn't know Liau Yung Siang reported that already.

Thanks for the review, it's already late In my timezone, but I'll update the PR tomorrow!

Commit:
Address PR review comments

Commit:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	doc/source/whatsnew/v0.23.4.txt

Commit:
Resolve merge conflict

Templarrr commented:
@jreback  I've confirmed that this solves the error that @YS-L reported in #17128 and addressed your PR review comments. Can you look again?

Also yesterday the 0.23.4 changelog got updated with today's date, but I don't see on pypi newer pandas version released yet, so I'm not entirely sure - will my bugfix be a part of 0.23.4 or should I move the comment to 0.24.0 changelog?

Templarrr commented:
The failure in travis is some kind of network glitch, unrelated to these changes :(

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Move changelog line to 24.0

Templarrr commented:
@jreback I see 0.23.4 was released, so this fix obviously didn't make it :) I've moved the changelog line to 0.24.0

Commit:
Merge remote-tracking branch 'upstream/master'

Templarrr commented:
@jreback I see you've moved the label to 0.23.5, didn't know there was 0.23.5 planned :) 
There is no 0.23.5 changelog currently, should I create a new one and move my changelog line there?

jreback commented:
yeah we might do a 0.23.5. if you want to push a 0.23.5 whats (empty) pls do , but a new PR pls.

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -607,6 +607,7 @@ Indexing
 - Fixed ``DataFrame[np.nan]`` when columns are non-unique (:issue:`21428`)
 - Bug when indexing :class:`DatetimeIndex` with nanosecond resolution dates and timezones (:issue:`11679`)
 - Bug where indexing with a Numpy array containing negative values would mutate the indexer (:issue:`21867`)
+- Fix ``OverflowError`` when trying to use 'inf' as ``na_value`` with integer index column (:issue:`17128`)
```
can you move to 0.23.5 (rebase on master to see the whatsnew)

Templarrr replied:
will do in few minutes

Templarrr replied:
@jreback  done.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Move changelog line to 23.5

jreback commented on doc/source/whatsnew/v0.23.5.txt:
```diff
@@ -40,3 +40,7 @@ Bug Fixes
 
 -
 -
+
```
can you put this in the IO section, and referencde `:func:read_csv` here

jreback reviewed (changes requested):
ping on green.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Modify changelog line from PR review feedback

jreback commented:
thanks @Templarrr

Templarrr commented:
You're welcome! Always glad to help :)
